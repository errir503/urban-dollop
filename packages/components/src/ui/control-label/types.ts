/**
 * Internal dependencies
 */
import type { Props as TextProps } from '../../text/types';

export type Props = TextProps & {
	isBlock?: boolean;
	size?: 'large' | 'medium' | 'small';
};
