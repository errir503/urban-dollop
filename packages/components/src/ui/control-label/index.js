export { default as ControlLabel } from './component';
export * from './hook';
