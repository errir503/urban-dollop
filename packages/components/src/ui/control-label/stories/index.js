/**
 * Internal dependencies
 */
import { ControlLabel } from '../index';

export default {
	component: ControlLabel,
	title: 'G2 Components (Experimental)/ControlLabel',
};

export const _default = () => {
	return <ControlLabel>Label</ControlLabel>;
};
