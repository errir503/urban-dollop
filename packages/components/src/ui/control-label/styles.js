/**
 * External dependencies
 */
import { css } from '@emotion/react';

/**
 * Internal dependencies
 */
import CONFIG from '../../utils/config-values';
import { getHighDpi } from '../utils/get-high-dpi';

const lineHeight = `calc(${ CONFIG.fontSize } * 1.2)`;

/**
 * @param {keyof CONFIG} size The padding size.
 */
function getPadding( size ) {
	return `calc((${ CONFIG[ size ] } - ${ lineHeight }) / 2)`;
}

const highDpiAdjust = getHighDpi`
	> * {
		position: relative;
		top: 0.5px;
	}
`;

export const ControlLabel = css`
	display: inline-block;
	line-height: ${ lineHeight };
	margin: 0;
	max-width: 100%;
	padding-bottom: ${ getPadding( 'controlHeight' ) };
	padding-top: ${ getPadding( 'controlHeight' ) };

	&:active {
		user-select: none;
	}

	${ highDpiAdjust };
`;

export const large = css`
	padding-bottom: ${ getPadding( 'controlHeightLarge' ) };
	padding-top: ${ getPadding( 'controlHeightLarge' ) };
`;

export const medium = css`
	padding-bottom: ${ getPadding( 'controlHeight' ) };
	padding-top: ${ getPadding( 'controlHeight' ) };
`;

export const small = css`
	padding-bottom: ${ getPadding( 'controlHeightSmall' ) };
	padding-top: ${ getPadding( 'controlHeightSmall' ) };
`;

export const inline = css`
	display: inline-block;
	vertical-align: middle;
`;

export const block = css`
	display: block;
`;
