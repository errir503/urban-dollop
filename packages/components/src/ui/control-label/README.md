# ControlLabel

<div class="callout callout-alert">
This feature is still experimental. “Experimental” means this is an early implementation subject to drastic and breaking changes.
</div>

`ControlLabel` is a form component that works with `FormGroup` to provide a label for form elements (e.g. `Switch` or `TextInput`).

## Usage

```jsx
import { ControlLabel, FormGroup, TextInput } from '@wordpress/components/ui';

function Example() {
	return (
		<FormGroup>
			<ControlLabel>First Name</ControlLabel>
			<TextInput />
		</FormGroup>
	);
}
```
