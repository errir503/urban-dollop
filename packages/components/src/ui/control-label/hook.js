/**
 * Internal dependencies
 */
import { useContextSystem } from '../context';
import { useFormGroupContextId } from '../form-group';
import { useText } from '../../text';
import * as styles from './styles';
import { useCx } from '../../utils/hooks/use-cx';

/**
 * @param {import('../context').WordPressComponentProps<import('./types').Props, 'label', false>} props
 */
export function useControlLabel( props ) {
	const {
		htmlFor: htmlForProp,
		isBlock = false,
		size = 'medium',
		truncate = true,
		...otherProps
	} = useContextSystem( props, 'ControlLabel' );

	const { className, ...textProps } = useText( {
		...otherProps,
		isBlock,
		truncate,
	} );

	const cx = useCx();

	const htmlFor = useFormGroupContextId( htmlForProp );
	const classes = cx(
		styles.ControlLabel,
		styles[ /** @type {'small' | 'medium' | 'large'} */ ( size ) ],
		className,
		isBlock ? styles.block : styles.inline
	);

	return {
		...textProps,
		className: classes,
		htmlFor,
	};
}
