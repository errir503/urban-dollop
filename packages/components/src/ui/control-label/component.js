/**
 * Internal dependencies
 */
import { contextConnect } from '../context';
import { View } from '../../view';
import { useControlLabel } from './hook';

/**
 * @param {import('../context').WordPressComponentProps<import('./types').Props, 'label', false>} props
 * @param {import('react').Ref<any>}                                                              forwardedRef
 */
function ControlLabel( props, forwardedRef ) {
	const controlLabelProps = useControlLabel( props );

	return <View as="label" { ...controlLabelProps } ref={ forwardedRef } />;
}

/**
 * `ControlLabel` is a form component that works with `FormGroup` to provide a
 * label for form elements (e.g. `Switch` or `TextInput`).
 *
 * ```jsx
 * import { ControlLabel, FormGroup, TextInput } from '@wordpress/components/ui';
 *
 * function Example() {
 * 	return (
 * 		<FormGroup>
 * 			<ControlLabel>First Name</ControlLabel>
 * 			<TextInput />
 * 		</FormGroup>
 * 	);
 * }
 * ```
 */
const ConnectedControlLabel = contextConnect( ControlLabel, 'ControlLabel' );

export default ConnectedControlLabel;
