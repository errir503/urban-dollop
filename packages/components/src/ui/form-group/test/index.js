/**
 * External dependencies
 */
import { render } from '@testing-library/react';

/**
 * Internal dependencies
 */
import { ControlLabel } from '../../control-label';
import { FormGroup, useFormGroupContextId } from '../index';

// @todo: Refactor this after adding next TextInput component.
const TextInput = ( { id: idProp, ...props } ) => {
	const id = useFormGroupContextId( idProp );
	return <input type="text" id={ id } { ...props } />;
};

// We're intentionally using a string literal for the ID to ensure
// the htmlFor and id properties of the label and inputs are bound correctly.
/* eslint-disable no-restricted-syntax */
describe( 'props', () => {
	test( 'should render correctly', () => {
		const { container } = render(
			<FormGroup id="fname" label="First name">
				<TextInput />
			</FormGroup>
		);

		const label = container.querySelector( 'label' );
		expect( label ).toHaveAttribute( 'for', 'fname' );
		expect( label ).toContainHTML( 'First name' );

		const input = container.querySelector( 'input' );
		expect( input ).toHaveAttribute( 'id', 'fname' );
	} );

	test( 'should render label without prop correctly', () => {
		const { container } = render(
			<FormGroup id="fname">
				<ControlLabel htmlFor="fname">First name</ControlLabel>
				<TextInput />
			</FormGroup>
		);

		const label = container.querySelector( 'label' );
		expect( label ).toHaveAttribute( 'for', 'fname' );
		expect( label ).toContainHTML( 'First name' );
	} );

	test( 'should render labelHidden', () => {
		const { container } = render(
			<FormGroup labelHidden id="fname" label="First name">
				<TextInput />
			</FormGroup>
		);

		const label = container.querySelector( 'label' );
		expect( label ).toContainHTML( 'First name' );
		// @todo: Refactor this after adding next VisuallyHidden.
		expect( label ).toHaveClass( 'components-visually-hidden' );
	} );

	test( 'should render alignLabel', () => {
		const { container } = render(
			<FormGroup alignLabel="right" id="fname" label="First name">
				<TextInput />
			</FormGroup>
		);

		expect( container.firstChild ).toMatchSnapshot();
	} );

	test( 'should render vertically', () => {
		const { container } = render(
			<FormGroup horizontal={ false } id="fname" label="First name">
				<TextInput />
			</FormGroup>
		);

		expect( container.firstChild ).toMatchSnapshot();
	} );
} );
/* eslint-enable no-restricted-syntax */
