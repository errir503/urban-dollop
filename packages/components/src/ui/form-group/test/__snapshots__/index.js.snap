// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render alignLabel 1`] = `
.emotion-0 {
  width: 100%;
}

.emotion-2 {
  display: inline-block;
  line-height: calc(13px * 1.2);
  margin: 0;
  max-width: 100%;
  padding-bottom: calc((36px - calc(13px * 1.2)) / 2);
  padding-top: calc((36px - calc(13px * 1.2)) / 2);
  padding-bottom: calc((36px - calc(13px * 1.2)) / 2);
  padding-top: calc((36px - calc(13px * 1.2)) / 2);
  color: #1e1e1e;
  line-height: 1.2;
  margin: 0;
  font-size: calc((13 / 13) * 13px);
  font-weight: normal;
  text-align: right;
  display: inline-block;
  vertical-align: middle;
}

.emotion-2:active {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

@media ( -webkit-min-device-pixel-ratio: 1.25 ), ( min-resolution: 120dpi ) {
  .emotion-2>* {
    position: relative;
    top: 0.5px;
  }
}

<div
  class="components-form-group emotion-0 emotion-1"
  data-wp-c16t="true"
  data-wp-component="FormGroup"
>
  <label
    class="components-truncate components-text components-control-label emotion-2 emotion-1"
    data-wp-c16t="true"
    data-wp-component="ControlLabel"
    for="fname"
  >
    First name
  </label>
  <input
    id="fname"
    type="text"
  />
</div>
`;

exports[`props should render vertically 1`] = `
.emotion-0 {
  width: 100%;
}

.emotion-2 {
  display: inline-block;
  line-height: calc(13px * 1.2);
  margin: 0;
  max-width: 100%;
  padding-bottom: calc((36px - calc(13px * 1.2)) / 2);
  padding-top: calc((36px - calc(13px * 1.2)) / 2);
  padding-bottom: calc((36px - calc(13px * 1.2)) / 2);
  padding-top: calc((36px - calc(13px * 1.2)) / 2);
  color: #1e1e1e;
  line-height: 1.2;
  margin: 0;
  font-size: calc((13 / 13) * 13px);
  font-weight: normal;
  text-align: left;
  display: inline-block;
  vertical-align: middle;
}

.emotion-2:active {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

@media ( -webkit-min-device-pixel-ratio: 1.25 ), ( min-resolution: 120dpi ) {
  .emotion-2>* {
    position: relative;
    top: 0.5px;
  }
}

<div
  class="components-form-group emotion-0 emotion-1"
  data-wp-c16t="true"
  data-wp-component="FormGroup"
>
  <label
    class="components-truncate components-text components-control-label emotion-2 emotion-1"
    data-wp-c16t="true"
    data-wp-component="ControlLabel"
    for="fname"
  >
    First name
  </label>
  <input
    id="fname"
    type="text"
  />
</div>
`;
